# Lab 3

### Objectives

1. Understand how to use Pytorch to write a Neural Network
2. Write a neural network with multiple hidden layers

In [1]:
import torch
from torchvision import transforms, datasets
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

Importing data.

We are using the MNIST database: http://yann.lecun.com/exdb/mnist/.
- The input is a 28 by 28 matrix (2-dimensional tensor), which corresponds to the
pixels of the image.
- The output is an integer between 0 and 9.

In [2]:
train = datasets.MNIST(
    "", 
    train=True, 
    download=True, 
    transform=transforms.Compose([transforms.ToTensor()])) #compose allows you to do multiple transformations, .ToTensor() converts a PIL Image or numpy.ndarray to tensor.

test = datasets.MNIST("", train=False, download=True, transform=transforms.Compose([transforms.ToTensor()]))

# DataLoader() wraps an iterable over the given dataset and supports automatic batching, sampling, shuffling and multiprocess data loading
trainset = torch.utils.data.DataLoader(train, batch_size=10, shuffle=True)
testset = torch.utils.data.DataLoader(test, batch_size=10, shuffle=True)

Defining the network.

In [3]:
class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(28*28, 64) # linear transformation, 𝑥𝐖+𝑏xW+b, 28*28 inputs, 64 outputs
        self.fc2 = nn.Linear(64, 64) # automatically creates the weight and bias tensors
        self.fc3 = nn.Linear(64, 64)
        self.fc4 = nn.Linear(64, 10)

    def forward(self, x):
        x = torch.sigmoid(self.fc1(x)) # passing the linearly transformed input through a neuron
        x = torch.sigmoid(self.fc2(x))
        x = torch.sigmoid(self.fc3(x))
        x = self.fc4(x)
        return F.softmax(x, dim=1)

net = Net()

Setting up the optimizer, the error function, and compute the gradient wrt the loss and then train for number of epochs.

In [4]:
optimizer = optim.SGD(net.parameters(), lr=0.001)

for epoch in range(3):
    for data in trainset:
        X, y = data
        net.zero_grad()
        output = net.forward(X.view(-1, 28*28))
        loss = F.nll_loss(output, y)
        loss.backward()
        optimizer.step()

Train network and test.

In [5]:
correct =0
total = 0

with torch.no_grad():
    for data in testset:
        X, y = data
        output = net.forward(X.view(-1, 28*28))
        for idx, i in enumerate(output):
            if torch.argmax(i) == y[idx]:
                correct += 1
            total += 1

print("Accuracy: ", round(correct/total, 3))

Accuracy:  0.098


This network produces an accuracy of 0.098. Why is this so low?


**Can you see a significant change in the performance of the network? Why do you think that’s the case?** </br>
The accuracy seems to be either 0.9 and above or around 0.1 - such extremes!!